lib/pbench/server/api/resources/metadata_api.py

```diff
@@ -0,0 +1,355 @@
+        except Exception:
+            self.logger.exception(
+                "Exception occurred while querying the Metadata model"
+            )
+            abort(500, message="INTERNAL ERROR")
+
+        response_object = {
+            "message": "success",
+            "data": {"sessions": data},
+        }
+        return make_response(jsonify(response_object), 200)
+
+
+class QueryMetadata(Resource):
+    """
+    Abstracted pbench API for querying a single user metadata session
+    """
+
+    def __init__(self, config, logger):
+        self.server_config = config
+        self.logger = logger
+
+    def verify_metadata(self, session, action):
+        current_user_id = Auth.token_auth.current_user().id
+        metadata_user_id = session.user_id
+        if current_user_id != metadata_user_id:
+            self.logger.warning(
+                "Metadata {}: Logged in user_id {} is different than the one provided in the URI {}",
+                action,
+                current_user_id,
+                metadata_user_id,
+            )
+            abort(403, message="Not authorized to perform the specified action")
+
+    @Auth.token_auth.login_required()
+    def get(self, id=None):
+        """
+        Get request for querying a metadata session of a user given a metadata id.
+        This requires a Pbench auth token in the header field
+
+        The url requires a metadata session id such as /user/metadata/<int:id>
+
+        Required headers include
+            Authorization:   Bearer <Pbench authentication token (user received upon login)>
+
+        :return: JSON Payload
+            response_object = {
+                    "message": "success",
+                    "data": {
+                        "id": "metadata_id",
+                        "value": "client text blob"
+                        "created": "Datetime string"
+                        "updated": "Datetime String"
+                        "key": "key string"
+                        }
+                }
+        """
+        if not id:
+            self.logger.warning("Metadata id not provided during metadata query")
+            abort(400, message="Please provide a metadata id to query")
+
+        try:
+            # Fetch the metadata session
+            metadata_session = Metadata.query(id=id)[0]
+        except Exception:
+            self.logger.exception(
+                "Exception occurred in the GET request while querying the Metadata model, id: {}",
+                id,
+            )
+            abort(500, message="INTERNAL ERROR")
+
+        # Verify if the metadata session id in the url belongs to the logged in user
+        self.verify_metadata(metadata_session, "get")
+
+        response_object = {
+            "message": "success",
+            "data": metadata_session.get_json(),
+        }
+        return make_response(jsonify(response_object), 200)
+
+    @Auth.token_auth.login_required()
+    def put(self, id=None):
```

Comment:
Should we allow public metadata objects to be editable?

Reply:
It's an interesting question, and I don't think that there is a clear-cut answer, but, given that public metadata objects can be freely created, I frankly have trouble objecting to allowing them to be publicly editable.

My suggestion for ameliorating the situation is to generate object IDs using random hashes (or UUIDs) instead of using sequential integers and to disable the API which returns the list of existing IDs (at least for the non-logged-in case).  This means that a client can create a metadata object, but, if they want to do anything with it, they have to remember the ID; and, given an ID, they are free to update or delete it (so long as the auth token matches or isn't present), as there is no way to list the existing IDs and they should be difficult to guess.

Reply:
Right; if non-login users can create metadata, but can't delete them or modify them, that's going to be a problem.

Reply:
It's not a problem for the client, Dave -- instead of modifying it, it just creates a new one!  :-D

Reply:
fixed